multicore_runtime/wsgi_config.py

```diff
@@ -0,0 +1,208 @@
+from google.appengine.api import appinfo_includes
+from google.appengine.runtime import wsgi
+
+
+def get_module_config_filename():
+  """Returns the name of the module configuration file (app.yaml)."""
+  module_yaml_path = os.environ['MODULE_YAML_PATH']
+  logging.info('Using module_yaml_path from env: %s', module_yaml_path)
+  return module_yaml_path
+
+
+def get_module_config(filename):
+  """Returns the parsed module config."""
+  with open(filename) as f:
+    return appinfo_includes.Parse(f)
+
+
+def app_for_script(script):
+  """Returns the WSGI app specified in the input string, or None on failure."""
+  try:
+    app, filename, err = wsgi.LoadObject(script)  # pylint: disable=unused-variable
+  except ImportError as e:
+    # Despite nominally returning an error object, LoadObject will sometimes
+    # just result in an exception. Since we're already processing the err
+    # object, we might as well just use that variable to store the exception.
+    err = e
+  if err:
+    # Log the exception but do not reraise.
+    logging.exception('Failed to import %s: %s', script, err)
+    return None
+  else:
+    return app_wrapped_in_user_middleware(app)
+
+
+def app_wrapped_in_user_middleware(app):
+  """Returns the input WSGI app, wrapped in appengine_config middleware."""
+  add_middleware = get_add_middleware_from_appengine_config()
+  if add_middleware:
+    return add_middleware(app)
+  else:
+    return app
+
+
+def get_add_middleware_from_appengine_config():
+  """Tries to import appengine_config and return middleware; fails silently.
+
+  `env_config` is optionally part of GAE-compatible user code.
+  If the user chooses to define the `webapp_add_wsgi_middleware` function
+  there, then we will use it to wrap app scripts. This is used for e.g.
+  appstats. Many user apps do not have or need env_config, so failure
+  to import it or failure to find `webapp_add_wsgi_middleware` is not an
+  error.
+
+  Returns:
+    The appengine_config.webapp_add_wsgi_middleware function or None.
+  """
+  try:
+    import appengine_config  # pylint: disable=g-import-not-at-top
+    try:
+      return appengine_config.webapp_add_wsgi_middleware
+    except AttributeError:
+      return None
+  except ImportError:
+    return None
+
+
+def static_app_for_handler(handler):
+  """Returns a WSGI app that serves static files as directed by the handler.
+
+  Args:
+    handler: An individual handler from appinfo_external.handlers
+      (appinfo.URLMap)
+
+  Returns:
+    A static file-serving WSGI app closed over the handler information.
+  """
+  regex = handler.url
+  files = handler.static_files
+  upload = handler.upload
+  if not files:
+    if handler.static_dir:
+      # If static_files is not set, convert static_dir to static_files and also
+      # modify the url regex accordingly. See the appinfo.URLMap docstring for
+      # more information.
+      regex = static_dir_url(handler)
+      files = handler.static_dir + r'/\1'
```

Comment:
See previous comments about using `%` and `.format()` instead of `+` for string concatenation.

Though perhaps in this case you might want to use `os.path.join` or `posixpath.join` since it will detect cases such as `static_dir` already ending in a slash?